Bad Kubernetes Secret Breaks ArgoCD Deployment

Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [X] I've included steps to reproduce the bug.
* [X] I've pasted the output of `argocd version`.

**Describe the bug**

I added a Secret to ArgoCD when toying with ApplicationSets and it ended up bricking the ArgoCD Deployment.

Hopefully this is enough to reproduce the issue - I ran through it twice with the instructions as is and I could recreate the situation both times.

**To Reproduce**

1. Clone https://github.com/poblahblahblah/argocd-test

2. Set up cluster via kind, create namespaces, install argocd and applicationsets:

```
git checkout 805771a886b20314c93ca87f4bfeabbecd380366

# this command creates a kind cluster, creates the namespaces, then installs argocd
./create_dev_env.sh
```

3. Create initial argocd Application:

```
git checkout bc9f0aa86c3d959b549f51e29ecf1d76c8138d91
kubectl apply -f apps.yaml
```

4. Add bad secret. This creates a "bad" secret which causes the argocd-application-controller and argocd-server to crashloop.

```
git checkout 20ed82c0268e5a2254962101d5ea477fac8a6823
kubectl apply -f apps.yaml
```

5. Try to repair by pointing it to a fixed SHA and manually deleting the secret:

```
git checkout 667cd321328efb021a2bf55461ccde4930c2e07e
kubectl apply -f apps.yaml
kubectl -n argocd delete secret argocd-kind
```

6. Monitor secrets to make sure ArgoCD doesn't keep creating the secret. I had to repeatedly delete the secret since ArgoCD kept recreating it, even when pointing to a known good sha:

```
watch 'kubectl -n argocd get secrets'
```

At this point the UI is down and the CLI can't initiate a login presumably due to the argocd-application-controller-0 pod being in a crashloop.

**Expected behavior**

I would expect that ArgoCD processes and UI would throw an error, but not cause a crashloop.

**Screenshots**

Crashloop:

```
NAME                                                READY   STATUS             RESTARTS   AGE
argocd-application-controller-0                     0/1     CrashLoopBackOff   7          23m
argocd-applicationset-controller-76fc4dc766-56krh   1/1     Running            0          23m
argocd-dex-server-9dc558f5-srq5c                    1/1     Running            0          23m
argocd-redis-759b6bc7f4-7vnnr                       1/1     Running            0          23m
argocd-repo-server-5fbf484547-4nlb2                 1/1     Running            0          23m
argocd-server-6d4678f7f6-5k4d9                      1/1     Running            4          23m
```


**Version**

```shell
~ ❯ argocd version
argocd: v2.0.3+8d2b13d.dirty
  BuildDate: 2021-05-27T19:03:08Z
  GitCommit: 8d2b13d733e1dff7d1ad2c110ed31be4804406e2
  GitTreeState: dirty
  GoVersion: go1.16.4
  Compiler: gc
  Platform: darwin/amd64
```

**Logs**

pod crashloop log:
```
time="2021-06-08T00:15:56Z" level=info msg="Processing all cluster shards"
time="2021-06-08T00:15:56Z" level=info msg="appResyncPeriod=3m0s"
time="2021-06-08T00:15:56Z" level=info msg="Application Controller (version: v2.0.2+9a7b0bc, built: 2021-05-20T19:30:25Z) starting (namespace: argocd)"
time="2021-06-08T00:15:56Z" level=info msg="Starting configmap/secret informers"
time="2021-06-08T00:15:56Z" level=info msg="Configmap/secret informer synced"
time="2021-06-08T00:15:56Z" level=info msg="Ignore status for CustomResourceDefinitions"
time="2021-06-08T00:15:56Z" level=info msg="0xc0008b62a0 subscribed to settings updates"
time="2021-06-08T00:15:56Z" level=info msg="Refreshing app status (comparison expired. reconciledAt: 2021-06-08 00:10:46 +0000 UTC, expiry: 3m0s), level (2)" application=my-application
time="2021-06-08T00:15:56Z" level=info msg="Refreshing app status (comparison expired. reconciledAt: 2021-06-08 00:10:46 +0000 UTC, expiry: 3m0s), level (2)" application=devops-toolkit
time="2021-06-08T00:15:56Z" level=info msg="Starting clusterSecretInformer informers"
time="2021-06-08T00:15:56Z" level=info msg="Ignore status for CustomResourceDefinitions"
time="2021-06-08T00:15:56Z" level=info msg="Comparing app state (cluster: https://kubernetes.default.svc, namespace: production)" application=my-application
time="2021-06-08T00:15:56Z" level=info msg="Start syncing cluster" server="https://kubernetes.default.svc"
time="2021-06-08T00:15:56Z" level=info msg="Ignore status for CustomResourceDefinitions"
time="2021-06-08T00:15:56Z" level=info msg="Comparing app state (cluster: https://kubernetes.default.svc, namespace: production)" application=devops-toolkit
time="2021-06-08T00:15:56Z" level=info msg="Resuming in-progress operation. phase: Running, message: " application=my-application
time="2021-06-08T00:15:56Z" level=info msg="Ignore status for CustomResourceDefinitions"
time="2021-06-08T00:15:56Z" level=info msg="Comparing app state (cluster: https://kubernetes.default.svc, namespace: production)" application=my-application
W0608 00:15:56.577254       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
W0608 00:15:56.629921       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-06-08T00:15:56Z" level=info msg="getRepoObjs stats" application=my-application build_options_ms=0 helm_ms=0 plugins_ms=0 repo_ms=0 time_ms=231 unmarshal_ms=6 version_ms=224
time="2021-06-08T00:15:56Z" level=info msg="getRepoObjs stats" application=my-application build_options_ms=0 helm_ms=0 plugins_ms=0 repo_ms=0 time_ms=229 unmarshal_ms=8 version_ms=221
time="2021-06-08T00:15:56Z" level=info msg="Skipping auto-sync: another operation is in progress" application=my-application
time="2021-06-08T00:15:56Z" level=info msg="Ignore status for CustomResourceDefinitions"
time="2021-06-08T00:15:56Z" level=info msg=Syncing application=my-application skipHooks=false started=false syncId=00001-dIpJY
time="2021-06-08T00:15:56Z" level=info msg="Update successful" application=my-application
time="2021-06-08T00:15:56Z" level=info msg="Reconciliation completed" application=my-application dedup_ms=0 dest-name= dest-namespace=production dest-server="https://kubernetes.default.svc" diff_ms=2 fields.level=2 git_ms=231 health_ms=0 live_ms=0 settings_ms=0 sync_ms=0 time_ms=252
time="2021-06-08T00:15:56Z" level=info msg="Applying resource Secret/argocd-kind in cluster: https://192.168.202.1:443, namespace: argocd"
time="2021-06-08T00:15:57Z" level=info msg="Refreshing app status (controller refresh requested), level (1)" application=my-application
time="2021-06-08T00:15:57Z" level=info msg="Ignore status for CustomResourceDefinitions"
time="2021-06-08T00:15:57Z" level=info msg="Comparing app state (cluster: https://kubernetes.default.svc, namespace: production)" application=my-application
E0608 00:15:57.194498       1 runtime.go:78] Observed a panic: &json.SyntaxError{msg:"invalid character '\\'' looking for beginning of object key string", Offset:2} (invalid character '\'' looking for beginning of object key string)
goroutine 188 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x2528ee0, 0xc000b92450)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/runtime/runtime.go:74 +0x95
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/runtime/runtime.go:48 +0x86
panic(0x2528ee0, 0xc000b92450)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/argoproj/argo-cd/v2/util/db.secretToCluster(0xc000294640, 0xc0004a3408)
	/go/src/github.com/argoproj/argo-cd/util/db/cluster.go:321 +0x997
github.com/argoproj/argo-cd/v2/util/db.(*db).WatchClusters.func2(0x28997c0, 0xc000294640)
	/go/src/github.com/argoproj/argo-cd/util/db/cluster.go:155 +0x72
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)
	/go/pkg/mod/k8s.io/client-go@v0.20.4/tools/cache/controller.go:231
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/go/pkg/mod/k8s.io/client-go@v0.20.4/tools/cache/shared_informer.go:777 +0xc2
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000d2a760)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00088ff60, 0x2d1afa0, 0xc0004a04e0, 0x251d501, 0xc00013e360)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:156 +0x9b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000d2a760, 0x3b9aca00, 0x0, 0x1370801, 0xc00013e360)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:90
k8s.io/client-go/tools/cache.(*processorListener).run(0xc0004bc280)
	/go/pkg/mod/k8s.io/client-go@v0.20.4/tools/cache/shared_informer.go:771 +0x95
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc0006f2300, 0xc0009e8560)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:73 +0x51
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:71 +0x65
panic: invalid character '\'' looking for beginning of object key string [recovered]
	panic: invalid character '\'' looking for beginning of object key string

goroutine 188 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/runtime/runtime.go:55 +0x109
panic(0x2528ee0, 0xc000b92450)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/argoproj/argo-cd/v2/util/db.secretToCluster(0xc000294640, 0xc0004a3408)
	/go/src/github.com/argoproj/argo-cd/util/db/cluster.go:321 +0x997
github.com/argoproj/argo-cd/v2/util/db.(*db).WatchClusters.func2(0x28997c0, 0xc000294640)
	/go/src/github.com/argoproj/argo-cd/util/db/cluster.go:155 +0x72
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)
	/go/pkg/mod/k8s.io/client-go@v0.20.4/tools/cache/controller.go:231
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/go/pkg/mod/k8s.io/client-go@v0.20.4/tools/cache/shared_informer.go:777 +0xc2
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000d2a760)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0009bdf60, 0x2d1afa0, 0xc0004a04e0, 0x251d501, 0xc00013e360)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:156 +0x9b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000d2a760, 0x3b9aca00, 0x0, 0x1370801, 0xc00013e360)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:90
k8s.io/client-go/tools/cache.(*processorListener).run(0xc0004bc280)
	/go/pkg/mod/k8s.io/client-go@v0.20.4/tools/cache/shared_informer.go:771 +0x95
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc0006f2300, 0xc0009e8560)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:73 +0x51
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:71 +0x65
```

Bad secret being deleted repeatedly even after apps.yaml was updated to reference a "good" ref:

```
~/src/scratch/argocd-test @667cd321 ❯ kubectl -n argocd delete secret argocd-kind
secret "argocd-kind" deleted

~/src/scratch/argocd-test @667cd321 ❯ sleep 300

~/src/scratch/argocd-test @667cd321 5s ❯ kubectl -n argocd get secrets
NAME                                           TYPE                                  DATA   AGE
argocd-application-controller-token-d9dhk      kubernetes.io/service-account-token   3      26m
argocd-applicationset-controller-token-5fj7r   kubernetes.io/service-account-token   3      26m
argocd-dex-server-token-dgz7z                  kubernetes.io/service-account-token   3      26m
argocd-initial-admin-secret                    Opaque                                1      22m
argocd-kind                                    Opaque                                3      41s
argocd-redis-token-nrctk                       kubernetes.io/service-account-token   3      26m
argocd-secret                                  Opaque                                5      26m
argocd-server-token-26j4l                      kubernetes.io/service-account-token   3      26m
default-token-cqvm5                            kubernetes.io/service-account-token   3      26m

~/src/scratch/argocd-test @667cd321 ❯ kubectl -n argocd delete secret argocd-kind
secret "argocd-kind" deleted

~/src/scratch/argocd-test @667cd321 ❯ sleep 200

~/src/scratch/argocd-test @667cd321 ❯ kubectl -n argocd get secrets
NAME                                           TYPE                                  DATA   AGE
argocd-application-controller-token-d9dhk      kubernetes.io/service-account-token   3      31m
argocd-applicationset-controller-token-5fj7r   kubernetes.io/service-account-token   3      31m
argocd-dex-server-token-dgz7z                  kubernetes.io/service-account-token   3      31m
argocd-initial-admin-secret                    Opaque                                1      26m
argocd-kind                                    Opaque                                3      1s
argocd-redis-token-nrctk                       kubernetes.io/service-account-token   3      31m
argocd-secret                                  Opaque                                5      31m
argocd-server-token-26j4l                      kubernetes.io/service-account-token   3      31m
default-token-cqvm5                            kubernetes.io/service-account-token   3      31m

~/src/scratch/argocd-test @667cd321 ❯ kubectl -n argocd delete secret argocd-kind
secret "argocd-kind" deleted
```

CLI login failure:

```
~ ❯ argocd login localhost:8080 --username admin
WARNING: server is not configured with TLS. Proceed (y/n)? y
Password:
FATA[0203] rpc error: code = Unavailable desc = transport is closing
```

curl failing (rather than a screenshot of Firefox failing):

```
~ ❯ curl https://localhost:8080 -D-
curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to localhost:8080
```

--- a/util/db/cluster.go
+++ b/util/db/cluster.go
@@ -83,15 +83,16 @@ func (db *db) ListClusters(ctx context.Context) (*appv1.ClusterList, error) {
 		return nil, err
 	}
 	clusterList := appv1.ClusterList{
-		Items: make([]appv1.Cluster, len(clusterSecrets)),
+		Items: make([]appv1.Cluster, 0),
 	}
 	hasInClusterCredentials := false
-	for i, clusterSecret := range clusterSecrets {
+	for _, clusterSecret := range clusterSecrets {
 		cluster, err := secretToCluster(clusterSecret)
 		if err != nil {
 			log.Errorf("could not unmarshal cluster secret %s", clusterSecret.Name)
+			continue
 		}
-		clusterList.Items[i] = *cluster
+		clusterList.Items = append(clusterList.Items, *cluster)
 		if cluster.Server == appv1.KubernetesInternalAPIServerAddr {
 			hasInClusterCredentials = true
 		}

--- a/util/db/cluster.go
+++ b/util/db/cluster.go
@@ -87,8 +87,11 @@ func (db *db) ListClusters(ctx context.Context) (*appv1.ClusterList, error) {
 	}
 	hasInClusterCredentials := false
 	for i, clusterSecret := range clusterSecrets {
-		cluster := *secretToCluster(clusterSecret)
-		clusterList.Items[i] = cluster
+		cluster, err := secretToCluster(clusterSecret)
+		if err != nil {
+			log.Errorf("could not unmarshal cluster secret: %v", err)
+		}
+		clusterList.Items[i] = *cluster
 		if cluster.Server == appv1.KubernetesInternalAPIServerAddr {
 			hasInClusterCredentials = true
 		}
@@ -127,7 +130,11 @@ func (db *db) CreateCluster(ctx context.Context, c *appv1.Cluster) (*appv1.Clust
 		}
 		return nil, err
 	}
-	return secretToCluster(clusterSecret), db.settingsMgr.ResyncInformers()
+	cluster, err := secretToCluster(clusterSecret)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "could not unmarshal cluster secret: %v", err)
+	}
+	return cluster, db.settingsMgr.ResyncInformers()
 }
 
 // ClusterEvent contains information about cluster event
@@ -152,7 +159,11 @@ func (db *db) WatchClusters(ctx context.Context,
 	clusterEventHandler := cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			if secretObj, ok := obj.(*apiv1.Secret); ok {
-				cluster := secretToCluster(secretObj)
+				cluster, err := secretToCluster(secretObj)
+				if err != nil {
+					log.Errorf("could not unmarshal cluster secret: %v", err)
+					return
+				}
 				if cluster.Server == appv1.KubernetesInternalAPIServerAddr {
 					// change local cluster event to modified or deleted, since it cannot be re-added or deleted
 					handleModEvent(localCls, cluster)
@@ -176,8 +187,16 @@ func (db *db) WatchClusters(ctx context.Context,
 		UpdateFunc: func(oldObj, newObj interface{}) {
 			if oldSecretObj, ok := oldObj.(*apiv1.Secret); ok {
 				if newSecretObj, ok := newObj.(*apiv1.Secret); ok {
-					oldCluster := secretToCluster(oldSecretObj)
-					newCluster := secretToCluster(newSecretObj)
+					oldCluster, err := secretToCluster(oldSecretObj)
+					if err != nil {
+						log.Errorf("could not unmarshal old cluster secret: %v", err)
+						return
+					}
+					newCluster, err := secretToCluster(newSecretObj)
+					if err != nil {
+						log.Errorf("could not unmarshal new cluster secret: %v", err)
+						return
+					}
 					if newCluster.Server == appv1.KubernetesInternalAPIServerAddr {
 						localCls = newCluster
 					}
@@ -205,7 +224,7 @@ func (db *db) getClusterSecret(server string) (*apiv1.Secret, error) {
 		return nil, err
 	}
 	for _, clusterSecret := range clusterSecrets {
-		if secretToCluster(clusterSecret).Server == strings.TrimRight(server, "/") {
+		if cluster, err := secretToCluster(clusterSecret); err == nil && cluster.Server == strings.TrimRight(server, "/") {
 			return clusterSecret, nil
 		}
 	}
@@ -222,7 +241,7 @@ func (db *db) GetCluster(ctx context.Context, server string) (*appv1.Cluster, er
 			return nil, err
 		}
 	}
-	return secretToCluster(clusterSecret), nil
+	return secretToCluster(clusterSecret)
 }
 
 // UpdateCluster updates a cluster
@@ -242,7 +261,12 @@ func (db *db) UpdateCluster(ctx context.Context, c *appv1.Cluster) (*appv1.Clust
 	if err != nil {
 		return nil, err
 	}
-	return secretToCluster(clusterSecret), db.settingsMgr.ResyncInformers()
+	cluster, err := secretToCluster(clusterSecret)
+	if err != nil {
+		log.Errorf("could not unmarshal cluster secret: %v", err)
+		return nil, err
+	}
+	return cluster, db.settingsMgr.ResyncInformers()
 }
 
 // Delete deletes a cluster by name
@@ -313,12 +337,12 @@ func clusterToSecret(c *appv1.Cluster, secret *apiv1.Secret) error {
 }
 
 // secretToCluster converts a secret into a Cluster object
-func secretToCluster(s *apiv1.Secret) *appv1.Cluster {
+func secretToCluster(s *apiv1.Secret) (*appv1.Cluster, error) {
 	var config appv1.ClusterConfig
 	if len(s.Data["config"]) > 0 {
 		err := json.Unmarshal(s.Data["config"], &config)
 		if err != nil {
-			panic(err)
+			return nil, err
 		}
 	}
 
@@ -354,5 +378,5 @@ func secretToCluster(s *apiv1.Secret) *appv1.Cluster {
 		RefreshRequestedAt: refreshRequestedAt,
 		Shard:              shard,
 	}
-	return &cluster
+	return &cluster, nil
 }

--- a/util/db/cluster.go
+++ b/util/db/cluster.go
@@ -89,7 +89,7 @@ func (db *db) ListClusters(ctx context.Context) (*appv1.ClusterList, error) {
 	for i, clusterSecret := range clusterSecrets {
 		cluster, err := secretToCluster(clusterSecret)
 		if err != nil {
-			log.Errorf("could not unmarshal cluster secret: %v", err)
+			log.Errorf("could not unmarshal cluster secret %s", clusterSecret.Name)
 		}
 		clusterList.Items[i] = *cluster
 		if cluster.Server == appv1.KubernetesInternalAPIServerAddr {
@@ -132,7 +132,7 @@ func (db *db) CreateCluster(ctx context.Context, c *appv1.Cluster) (*appv1.Clust
 	}
 	cluster, err := secretToCluster(clusterSecret)
 	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "could not unmarshal cluster secret: %v", err)
+		return nil, status.Errorf(codes.InvalidArgument, "could not unmarshal cluster secret %s", clusterSecret.Name)
 	}
 	return cluster, db.settingsMgr.ResyncInformers()
 }
@@ -161,7 +161,7 @@ func (db *db) WatchClusters(ctx context.Context,
 			if secretObj, ok := obj.(*apiv1.Secret); ok {
 				cluster, err := secretToCluster(secretObj)
 				if err != nil {
-					log.Errorf("could not unmarshal cluster secret: %v", err)
+					log.Errorf("could not unmarshal cluster secret %s", secretObj.Name)
 					return
 				}
 				if cluster.Server == appv1.KubernetesInternalAPIServerAddr {
@@ -189,12 +189,12 @@ func (db *db) WatchClusters(ctx context.Context,
 				if newSecretObj, ok := newObj.(*apiv1.Secret); ok {
 					oldCluster, err := secretToCluster(oldSecretObj)
 					if err != nil {
-						log.Errorf("could not unmarshal old cluster secret: %v", err)
+						log.Errorf("could not unmarshal cluster secret %s", oldSecretObj.Name)
 						return
 					}
 					newCluster, err := secretToCluster(newSecretObj)
 					if err != nil {
-						log.Errorf("could not unmarshal new cluster secret: %v", err)
+						log.Errorf("could not unmarshal cluster secret %s", newSecretObj.Name)
 						return
 					}
 					if newCluster.Server == appv1.KubernetesInternalAPIServerAddr {
@@ -263,7 +263,7 @@ func (db *db) UpdateCluster(ctx context.Context, c *appv1.Cluster) (*appv1.Clust
 	}
 	cluster, err := secretToCluster(clusterSecret)
 	if err != nil {
-		log.Errorf("could not unmarshal cluster secret: %v", err)
+		log.Errorf("could not unmarshal cluster secret %s", clusterSecret.Name)
 		return nil, err
 	}
 	return cluster, db.settingsMgr.ResyncInformers()

--- a/util/db/cluster_test.go
+++ b/util/db/cluster_test.go
@@ -241,4 +241,15 @@ func TestListClusters(t *testing.T) {
 		require.NoError(t, err)
 		assert.Len(t, clusters.Items, 2)
 	})
+
+	t.Run("Implicit in-cluster secret", func(t *testing.T) {
+		kubeclientset := fake.NewSimpleClientset(validSecret2)
+		settingsManager := settings.NewSettingsManager(context.Background(), kubeclientset, fakeNamespace)
+		db := NewDB(fakeNamespace, settingsManager, kubeclientset)
+
+		clusters, err := db.ListClusters(context.TODO())
+		require.NoError(t, err)
+		// ListClusters() should have added an implicit in-cluster secret to the list
+		assert.Len(t, clusters.Items, 2)
+	})
 }

--- a/util/db/cluster_test.go
+++ b/util/db/cluster_test.go
@@ -13,6 +13,7 @@ import (
 
 	"github.com/argoproj/argo-cd/v2/common"
 	"github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	appv1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
 	"github.com/argoproj/argo-cd/v2/util/settings"
 )
 
@@ -178,4 +179,66 @@ func runWatchTest(t *testing.T, db ArgoDB, actions []func(old *v1alpha1.Cluster,
 		assert.Fail(t, "Failed due to timeout")
 	}
 
+}
+
+func TestListClusters(t *testing.T) {
+	validSecret1 := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "mycluster1",
+			Namespace: fakeNamespace,
+			Labels: map[string]string{
+				common.LabelKeySecretType: common.LabelValueSecretTypeCluster,
+			},
+		},
+		Data: map[string][]byte{
+			"server": []byte(appv1.KubernetesInternalAPIServerAddr),
+			"name":   []byte("in-cluster"),
+		},
+	}
+
+	validSecret2 := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "mycluster2",
+			Namespace: fakeNamespace,
+			Labels: map[string]string{
+				common.LabelKeySecretType: common.LabelValueSecretTypeCluster,
+			},
+		},
+		Data: map[string][]byte{
+			"server": []byte("http://mycluster2"),
+			"name":   []byte("mycluster2"),
+		},
+	}
+
+	invalidSecret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "mycluster3",
+			Namespace: fakeNamespace,
+		},
+		Data: map[string][]byte{
+			"name":   []byte("test"),
+			"server": []byte("http://mycluster3"),
+			"config": []byte("{'tlsClientConfig':{'insecure':false}}"),
+		},
+	}
+
+	t.Run("Valid clusters", func(t *testing.T) {
+		kubeclientset := fake.NewSimpleClientset(validSecret1, validSecret2)
+		settingsManager := settings.NewSettingsManager(context.Background(), kubeclientset, fakeNamespace)
+		db := NewDB(fakeNamespace, settingsManager, kubeclientset)
+
+		clusters, err := db.ListClusters(context.TODO())
+		require.NoError(t, err)
+		assert.Len(t, clusters.Items, 2)
+	})
+
+	t.Run("Cluster list with invalid cluster", func(t *testing.T) {
+		kubeclientset := fake.NewSimpleClientset(validSecret1, validSecret2, invalidSecret)
+		settingsManager := settings.NewSettingsManager(context.Background(), kubeclientset, fakeNamespace)
+		db := NewDB(fakeNamespace, settingsManager, kubeclientset)
+
+		clusters, err := db.ListClusters(context.TODO())
+		require.NoError(t, err)
+		assert.Len(t, clusters.Items, 2)
+	})
 }

--- a/util/db/cluster_test.go
+++ b/util/db/cluster_test.go
@@ -6,6 +6,7 @@ import (
 	"time"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes/fake"
@@ -37,7 +38,8 @@ func Test_secretToCluster(t *testing.T) {
 			"config": []byte("{\"username\":\"foo\"}"),
 		},
 	}
-	cluster := secretToCluster(secret)
+	cluster, err := secretToCluster(secret)
+	require.NoError(t, err)
 	assert.Equal(t, *cluster, v1alpha1.Cluster{
 		Name:   "test",
 		Server: "http://mycluster",
@@ -58,11 +60,29 @@ func Test_secretToCluster_NoConfig(t *testing.T) {
 			"server": []byte("http://mycluster"),
 		},
 	}
-	cluster := secretToCluster(secret)
+	cluster, err := secretToCluster(secret)
+	assert.NoError(t, err)
 	assert.Equal(t, *cluster, v1alpha1.Cluster{
 		Name:   "test",
 		Server: "http://mycluster",
 	})
+}
+
+func Test_secretToCluster_InvalidConfig(t *testing.T) {
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "mycluster",
+			Namespace: fakeNamespace,
+		},
+		Data: map[string][]byte{
+			"name":   []byte("test"),
+			"server": []byte("http://mycluster"),
+			"config": []byte("{'tlsClientConfig':{'insecure':false}}"),
+		},
+	}
+	cluster, err := secretToCluster(secret)
+	require.Error(t, err)
+	assert.Nil(t, cluster)
 }
 
 func TestUpdateCluster(t *testing.T) {

